lib/model/recent_senders.dart

```diff
@@ -0,0 +1,148 @@
+    if (message is! StreamMessage) return;
+    final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+    ((streamSenders[streamId] ??= {})
+      [senderId] ??= MessageIdTracker()).add(messageId);
+    (((topicSenders[streamId] ??= {})[topic] ??= {})
+      [senderId] ??= MessageIdTracker()).add(messageId);
+  }
+
+  void handleDeleteMessageEvent(DeleteMessageEvent event, Map<int, Message> cachedMessages) {
+    if (event.messageType != MessageType.stream) return;
+
+    final messagesByUser = <int, QueueList<int>>{};
+    for (final id in event.messageIds) {
+      final message = cachedMessages[id] as StreamMessage?;
+      if (message == null) continue;
+      (messagesByUser[message.senderId] ??= QueueList()).add(id);
+    }
+
+    final DeleteMessageEvent(:streamId!, :topic!) = event;
+    for (final entry in messagesByUser.entries) {
+      final MapEntry(key: senderId, value: messages) = entry;
+
+      final sendersByStream = streamSenders[streamId];
+      final messagesBySenderInStream = sendersByStream?[senderId];
+      messagesBySenderInStream?.removeAll(messages);
+      if (messagesBySenderInStream?.maxId == null) sendersByStream?.remove(senderId);
+      if (sendersByStream?.isEmpty ?? false) streamSenders.remove(streamId);
+
+      final topicsByStream = topicSenders[streamId];
+      final sendersByTopic = topicsByStream?[topic];
+      final messagesBySenderInTopic = sendersByTopic?[senderId];
+      messagesBySenderInTopic?.removeAll(messages);
+      if (messagesBySenderInTopic?.maxId == null) sendersByTopic?.remove(senderId);
+      if (sendersByTopic?.isEmpty ?? false) topicsByStream?.remove(topic);
+      if (topicsByStream?.isEmpty ?? false) topicSenders.remove(streamId);
+    }
+  }
+}
+
+@visibleForTesting
+class MessageIdTracker {
+  /// A list of distinct message IDs, sorted ascendingly.
+  @visibleForTesting
+  QueueList<int> ids = QueueList();
+
+  /// The maximum id in the tracker list, or `null` if the list is empty.
+  int? get maxId => ids.lastOrNull;
+
+  /// Add the message ID to the tracker list at the proper place, if not present.
+  ///
+  /// Optimized, taking O(1) time for the case where that place is the end,
+  /// because that's the common case for a message that is received through
+  /// [PerAccountStore.handleEvent]. May take O(n) time in some rare cases.
+  void add(int id) {
+    if (ids.isEmpty || id > maxId!) {
+      ids.addLast(id);
+      return;
+    }
+    final i = lowerBound(ids, id);
+    if (i < ids.length && ids[i] == id) {
+      // The ID is already present. Nothing to do.
+      return;
+    }
+    ids.insert(i, id);
+  }
+
+  /// Add the messages IDs to the tracker list at the proper place, if not present.
+  ///
+  /// [newIds] should be sorted ascendingly.
+  void addAll(QueueList<int> newIds) {
+    if (ids.isEmpty) {
+      ids = newIds;
+      return;
+    }
+    ids = setUnion(ids, newIds);
+  }
+
+  void removeAll(QueueList<int> idsToRemove) {
```

Comment:
nit: this doesn't use that it's a QueueList, so might as well take a general List (and the caller might as well make plain default lists with `[]`, rather than QueueLists)

Reply:
The motivation was that we use `QueueList` internally as well as in `addAll`, so I thought passing a `QueueList` in `removeAll` would align with them. Anyway, changed it back to `List`. 🙂